pkg/nsx/services/securitypolicy/builder.go

```diff
@@ -598,22 +580,44 @@ func (service *SecurityPolicyService) buildRuleOutGroup(obj *v1alpha1.SecurityPo
 	return nsxRuleDstGroup, nsxRuleSrcGroupPath, nsxRuleDstGroupPath, nsxGroupShare, nil
 }
 
-func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int, createdFor string) string {
-	serializedBytes, _ := json.Marshal(rule)
-	ruleHash := fmt.Sprintf("%s", util.Sha1(string(serializedBytes)))
-	ruleIdxStr := fmt.Sprintf("%d", ruleIdx)
+func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, ruleIdx int, createdFor string) string {
+	ruleIndexHash := service.buildRuleHashString(&(obj.Spec.Rules[ruleIdx]))
+
 	if IsVPCEnabled(service) {
-		suffix := strings.Join([]string{ruleIdxStr, ruleHash}, common.ConnectorUnderline)
-		return util.GenerateIDByObjectWithSuffix(obj, suffix)
+		ruleIndexHash = service.buildLimitedRuleHashString(&(obj.Spec.Rules[ruleIdx]))
+		return util.GenerateIDByObjectWithSuffix(obj, ruleIndexHash)
 	}
+
 	prefix := common.SecurityPolicyPrefix
 	if createdFor == common.ResourceTypeNetworkPolicy {
 		prefix = common.NetworkPolicyPrefix
 	}
-	return util.GenerateID(fmt.Sprintf("%s", obj.UID), prefix, ruleHash, ruleIdxStr)
+	ruleIdxSuffix := fmt.Sprintf("%d", ruleIdx)
+	return util.GenerateID(fmt.Sprintf("%s", obj.UID), prefix, ruleIdxSuffix, ruleIndexHash)
```

Comment:
This is using a different order from the previous solution in function `util.GenerateID` with T1 solution, old way: obj.UID, prefix, ruleHash, ruleId; new way: obj.UID, prefix, ruleID, ruleHash.

Reply:
In fact, current change is a bug fix to align with v4.1.2 ID format. 
https://github.com/vmware-tanzu/nsx-operator/blob/v4.1.2/pkg/nsx/services/securitypolicy/builder.go#L513
The bug was introduced in xie's refactor patch:
https://github.com/vmware-tanzu/nsx-operator/commit/0ed01de3c89f89e6c9c2f9c02f59065572aa98bf#diff-878f5907665c91bbee664334052881facd7f44a8617ab92547f46a730c8a97d9R571